Add lucas tdf

erigler-usgs opened the pull request:
This is a 2nd attempt to review the add_lucas_Tdf branch and merge it into master. In my effort to shrink the size of the repo (>200 MB due to an unfortunate history of tracking changes to hundreds of digital images that did not need to be tracked), the size had grown to something rather ridiculous. The only fix I could figure out involved re-writing the history with those image removed. This then mangled references in the previous PR.

My apologies for the tedium of asking for a re-review. For reference, you should be able to see the old review here: https://github.com/usgs/geomag-imp/pull/12.

erigler-usgs committed:
Replace legacy algorithm in Make_T with Lucas' Tdf

The Make_T method is probably the biggest cpu-hog in this whole package.
Greg Lucas' pysecs re-implemented the math of Amm & Viljanen to be more
vectorized (Numpy-friendly), and use fewer trigonometry calls.

erigler-usgs committed:
Add test for Bphi

This test only works for Bphi if we place the SEC pole on the equator,
and predict Bphi at 90 degrees longitude away from this pole, and on
the equator. Basically, we're using the equation for Btheta from Amm
and Viljanen (1999), but rotated by 90 degrees.

greglucas reviewed (approved):
This is porting over my approving review from https://github.com/usgs/geomag-imp/pull/12
See the inline comments over there for minor suggestions.

> Looks good to me, I checked the repo out, built it and ran the tests locally that all passed.
> I am fine with you borrowing the code from pysecs. I am not sure if you've ported over the time axis vectorization yet or not, but that is another speed enhancement to keep in mind for the future as well.

erigler-usgs committed:
Remove erroneous comments from Lucas' original

Make_T is 90+% borrowed from Greg Lucas' pysecs package, but one of the
changes made for compatibility with Geomag-IMP was to output Btheta, Bphi,
and Brho, rather than Bx, By, and Bz. I removed a comment that related
to this.

erigler-usgs committed:
Remove superflous comment about coordinate frames

There was a comment describing the coordinate frames used in Greg Lucas'
pysecs package that was not actually very relevant to the calc_bearing
method. I removed this comment to avoid future confusion. calc_bearing
works exactly as intended with no changes.

erigler-usgs committed:
Removed left-over debug print statements.

erigler-usgs committed:
Add blank line to end of secs.py

PEP8 compliance

erigler-usgs committed:
Use class objects when available

I was re-calculating nsecs (number of SECs) inside make_T when this
value was already available within an initialized object.

erigler-usgs commented:
I am very thankful to @greglucas for his careful and helpful review of a substantial change to Geomag-IMP that has resulted in a very noticeable speedup when calculating the T transform matrix from local spherical to global geographic coordinates (a factor of more than 3X when interpolating ~14 observatories to over 2000 regular grid points using over 2000 SEC basis functions!).

To answer @greglucas's question about time axis vectorization, no, I did not attempt to port that functionality from pysecs. I took a different, and somewhat incompatible, approach to solving this problem in Geomag-IMP that I believe is a little better suited to real time operations. I have no doubt that his approach would be more efficient if batch processing large timeseries. Still, as noted above, just incorporating this one small part of pysecs (the T transform matrix generation) into Geomag-IMP led to a substantial speedup in a typical operating scenario.

I respond to all reviewer comments made in the old PR #12 below, and cite a commit used to address the issue when appropriate:

**Comment 1**
> ...this repo is HUGE, over 200 MB currently that took me a little while to download and clone locally.

Yes, yes it was. This was due to early and misguided commits of mostly images to the repository. I stopped this a long time ago, but didn't realize that when someone cloned the repo that they would receive everything ever committed. To remedy this, it was necessary to clean up and rewrite the Git history using the `git filter-branch` command. The repo is no ~20MB, which is still somewhat large, but the largest  of those files was necessary to run the Jupyter Notebook that was current at the time, so I didn't want to delete that.

**Comment 2**
> After our discussion, you may want to update this to say Btheta, Bphi, Br instead?

I've made an appropriate change, which can be found in commit be61e2fa6db3bc73aa9d216a4b2fbc87158a6cea

**Comment 3**
> Update this note too. This routine may have also added to the mix of coordinate discussions that we had yesterday.

The "note" you refer to is particularly relevant to the function referenced. The bearing is just an angle from one point in geographic coordinates to another point in geographic coordinates, which is why this function works for Geomag-IMP (i.e., Geomag-IMP's locations are still specified in {lat,lon,radius}, even if its magnetic field vectors are in {theta, phi, r}. I went ahead and deleted the inline comments here, since the docstring described accurately what was going on. This can be seen in commit 574fa1032023a14213e67f7d3787f1ca616dc34a.

**Comment 4**
> I suppose if it is easier for your code, you could avoid this opposite direction for bearing here so that your angles align directly with your coordinate frame

I don't think I want to change anything here. The calc_bearing method does exactly what any sane person would expect, so I just make a very minor adjustment to its output when I use it inside make_T.

**Comment 5**
> blank line at the end of the file

Thanks for the PEP8 reminder. Fixed in commit 70d58d5e7577832f858241b3405cb71fa55a393d.

**Comment 6**
> These should be automatically printed if there is a failure in the assertion below.

Thanks. This is now fixed in commit 13cf47051e596de772f58e6023d2f82a655873d4.

**Comment 7**
> In the previous code you had self.secs.n_secs, I assume these are equivalent, so depending on where you want the consistency you could change this over to that.

Good catch. This is fixed in commit 4538a7b45ea677aea1063fdf37001d75efe7191f.
